Adding Python Expression documentation

parmitam opened the pull request:
Addition to the documentation to add python expression documentation

parmitam committed:
Adding Python Expression documentation

BrandonHaynes commented on docs/myria-python/index.md:
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+            return 0
+    return 1
+
+MyriaPythonFunction(pyIsPrime, outType).register()
```
Why not a decorator?  Users shouldn't ever need to call `register` explicitly.

parmitam replied:
I haven't got the decorator to work, feel free to add it when you have it working correctly.thanks

BrandonHaynes replied:
Is there a GitHub issue with the problem you're encountering?  I know that @orzikhd has been able to use them.  Thanks.

orzikhd replied:
I just tried it out and was able to use the @myria_function decorator, but it should be noted that I'm on a slightly old version of Myria at this point. The following code worked:
```
from myria import *
%load_ext myria
connection = MyriaConnection(rest_url='http://localhost:8753')
%connect http://localhost:8753
from myria.udf import myria_function

@myria_function(name='bar', output_type="LONG_TYPE", connection=connection)
def bar(dt):
    import math
    n = dt[0][0]
    if n % 2 == 0 and n > 2: 
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return 0
    return 1
```

parmitam replied:
I was hoping just 
```python
@myria_function(output_type="LONG_TYPE")
def bar(dt):
    import math
    n = dt[0][0]
    if n % 2 == 0 and n > 2: 
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return 0
    return 1
```
 would work

parmitam replied:
Also, @BrandonHaynes  could we please update the demo cluster so these work on the demo cluster. 
a cluster deployed with myria-cluster and --unprovisioned flag works, so I am not sure what needs to happen.  
I would not be  comfortable pushing the doc changes if they don't work on the demo cluster. Thanks

BrandonHaynes replied:
If you set a default connection (`MyriaRelation.DefaultConnection = c`) then you don't have to specify it when registering.  If you haven't, myria-python has no idea which (and there may be >1) to use when connecting.

jortiz16 commented on docs/myria-python/index.md:
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+
+Myria allows arbitrary Python Functions as expressions which can be used in transformations like map, flatmap and aggregates.
+#### Registering a Python Function
+Python function needs to be registered before it can be used. Myria-python can be used to register python functions, list registered python functions as well as get details of the registered function.
```
Maybe reword this as... "Myria-Python can be used to register new Python functions, list existing registered Python functions, and to retrieve details about existing registered functions"

Also I did not see an example within this block where you list how to get details about the registered functions. You only show how to list them.

parmitam replied:
done.

jortiz16 commented on docs/myria-python/index.md:
````diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+connection.get_functions()
+```
+#### Invoking Python Expression in MyriaL
+A registered Python function can then be used in a MyriaL  query as an operator as:
````
"A registered Python function can be used in a MyriaL query as an operator:"

Is this an operator though? Isn't it an expression?

parmitam replied:
it is an expression, but it can be used as an operator, i'll simplify the language, i can see why this is confusing.

parmitam replied:
Done

jortiz16 commented on docs/myria-python/index.md:
````diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+```
+#### Python Expressions in UDAs
+Python functions call also be used in User Defined Aggregates in MyriaL.
+
````
Missing "can"

jortiz16 commented on docs/myria-python/index.md:
````diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+""", connection=connection)
+
+q.status
+```
````
Also, how can we run these? I wanted to check if all the examples work. Do I need to use a separate branch for anything via Myria-Cluster? Doesn't seem like the Myria Demo cluster can run these queries....

parmitam replied:
you need to deploy a cluster with  --unprovisioned.
@senderista  when will we have updated images for this?

senderista replied:
just for the record, we have updated images that can run the examples.

Commit:
code review changes

jortiz16 commented on docs/myria-python/index.md:
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+    tuplist = dt
+    state = None
+    for i in tuplist:
+        imgid = i12]
```
There is a syntax error here. After `i12` there shouldn't be a bracket

parmitam replied:
fixed.

jortiz16 commented on docs/myria-python/index.md:
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+            state[:,:,:,imgid] = img
+    return (state)
+
+MyriaPythonFunction(udfAgg, outType).register()
```
This should be`udfSum`

jortiz16 commented on docs/myria-python/index.md:
````diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+Python functions can also be used in User Defined Aggregates in MyriaL.
+
+```python
+#register a UDA
````
I think this should be `UDF`

Commit:
code review changes

BrandonHaynes commented on docs/myria-python/index.md:
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+connection.get_functions()
+
+#List details of a registered function
+connection.get_function('pyIsPrime')
```
myria-python has a higher-level API for getting functions (as entities, not JSON documents) as `MyriaFunction.get(name, connection)` and `MyriaFunction.get_all(connection)`.  There should be no for users to ever interact directly with the MyriaConnection facade.

I'll clean this up after I update the demo cluster.

jingjingwang committed:
Update 2015-05-03-cloud_service.md

jingjingwang committed:
Update 2015-05-03-cloud_service.md

BrandonHaynes committed:
Cleaned up UDF documentation

BrandonHaynes committed:
Merge branch 'parmitam-patch-1' of github.com:uwescience/myria-website into parmitam-patch-1

BrandonHaynes commented on docs/myria-python/index.md:
```diff
@@ -178,6 +178,95 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
+    --emit
+    [tm];
+  };
+  t = scan(public:adhoc:raw);
```
This relation doesn't exist on a vanilla Myria install (or on demo) -- need to stage the content on S3 and do an explicit load so that the query works.

BrandonHaynes commented:
UDF documentation is corrected.  I can remove two additional explicit connection references when https://github.com/uwescience/myria-python/pull/73 is merged.

The relation `public:adhoc:raw` needs to be explicitly loaded from a stable external source so that this works on a vanilla Myria install.

BrandonHaynes committed:
Use default connection instances

senderista commented:
Is this PR only blocking on our locating the data for the relation `public:adhoc:raw`? @parmitam can you address this?

parmitam committed:
Update index.md

changed the UDA to work on twitterK table instead, image data for raw is  copyrighted and cannot be loaded on public cluster.

parmitam commented:
Made a change to the UDA so we don't need the image relation, it works on twitterK relation, the UDA is trivial, but it gets the point across.
